### PR TITLE
Devops/2337 deploy develop on pr merge

### LINKED ISSUE #2337: [DevOps] Deploy `develop` to a cloudgov environment after PR merge
**Description:**
After PRs merge we should
* build and deploy the application/services to a development environment (bonus points for storing the build artifacts for later deployment to production)
* run end-to-end tests against the deployed application

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] `develop` deploys to [a cloudgov environment] when PRs merge
- [x] cypress tests run against [the cloudgov environment] once a deployment completes
    - [x] run tests after deployment to any environment? dev, staging, qasp, etc (not prod)
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Modify the Circle CI config to deploy the `develop` branch to [a development environment in cloudgov] after PRs merge
- [x] Add a step post-deployment to run the cypress e2e tests against [the environment]. 
   - This requires overriding some environment variables in the `cypress run` command, as well as adding a `CYPRESS_TOKEN` environment variable to the Circle CI config and CloudGov app variables.
   - example ci job: https://github.com/raft-tech/TANF-app/blob/be8197a46585ae006bd2e79a655326b302933157/.circleci/build-and-test/jobs.yml#L44-L65
   - example of overriding env vars in `cypress run`: https://github.com/raft-tech/TANF-app/blob/be8197a46585ae006bd2e79a655326b302933157/tdrs-frontend/package.json#L41
   - [update with link to example config once cypress PR merges]
- [x] Update devops/application documentation to reflect the change in process
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by George-Hudson**
## Summary of Changes
_This tracks pushes to the `develop` branch and triggers a build and deploy to the develop environment, followed by kicking off Cypress end to end integration testing against the newly deployed env_
Pull request closes #2337 

## How to Test
_This will set up a deployment pipeline that only tracks the develop branch_
Future tests for this will be when a PR merge (Including this one) kicks off a build, deploy, and test pipeline.
To test the functionality prior to this PR, instead of tracking the `develop` branch, this `devops/2337-deploy-develop-on-pr-merge` was put in as a placeholder, but still deployed to the develop environment.

This [GitHub Action](https://github.com/raft-tech/TANF-app/actions/runs/4125507980) kicked off the [CircleCI Pipeline](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/12444). Tests against the newly deployed environment can be seen as part of [that pipeline](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/12444/workflows/cb76f5b1-ab78-43c7-8e1f-2569c80a104f/jobs/34374?invite=true#step-109-39)

The tracked branch was changed to `develop` before initiating this PR. The deploy and e2e tests were not [kicked off](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/12446) but when this PR is merged, a corresponding GitHub Action and CircleCI build should immediately start.

## Deliverables
_This is part of a larger initiative to refine Continuous Integration and Continuous Deployment pipelines to improve developer quality of life, confidence in higher level deployments (staging and production), and enabling safe, secure, and granular updates to the live assets_

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Upon any merge to `develop` GitHub Actions kicks off a CircleCI pipeline
+ [x] That CircleCI pipeline deploys the changes to the develop environment
+ [x] After the deploy, Cypress Tests are run against the deployed environment
+ [x] The test results can be seen within the logs of the pipeline
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

This PR does not include frontend or backend changes and the app should function identically as before. The process should be continuously monitored by interested parties to look for improvements or breaking pipelines.

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

This PR used existing tools and scripts and no research was done.
Future research should look at creating:
- a stateful infrastructure manifest
- artifact storage
- environment variable and secrets storage
- CI/CD pipeline simplification

### CODE REVIEW

File: tdrs-frontend/package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "test:accessibility": "concurrently -k -s first 'REACT_APP_PA11Y_TEST=true npm run start:ci' 'wait-on http://localhost:3000/ && npm run pa11y-ci'",
     "test:e2e": "cypress open",
     "test:e2e-ci": "cypress run --headless -b chrome --env cypressToken=${CYPRESS_TOKEN}",
+    "test:e2e-develop": "cypress run --headless -b chrome --config baseUrl=https://tdp-frontend-develop.app.cloud.gov --env cypressToken=${CYPRESS_TOKEN},apiUrl=https://tdp-backend-develop.app.cloud.gov/v1",
```
Comment (andrew-jameson) [inline tdrs-frontend/package.json]: Hmm, this is going to need updating almost immediately. I'm sure it is functional now but we have two changes incoming:

1. Nginx Config #1544, all traffic will be going through frontend
2. New staging/develop FQDNs #2385

File: scripts/deploy-backend.sh [hunk 2]
```diff
@@ -42,7 +42,9 @@ set_cf_envs()
   "AV_SCAN_URL"
   "BASE_URL"
   "CLAMAV_NEEDED"
+  "CYPRESS_TOKEN"
   "DJANGO_CONFIGURATION"
+  "DJANGO_DEBUG"
```
Comment (andrew-jameson) [inline scripts/deploy-backend.sh:47]: Were these added to CircleCI env vars? Having missing/empty env vars has given us problems in the past.

File: docs/Technical-Documentation/circle-ci.md [hunk 3]
```diff
@@ -101,3 +101,18 @@ These all have defaults set in their respective settings modules, but may be ove
         * Bind the backend application to the S3 and RDS services in Cloud.gov
         * Run `/scripts/set-backend-env-vars.sh` (detailed above)
         * Restage the application to make environment variable and bound services live.
+
+## CI/CD Pipeline
+
```
Comment (andrew-jameson) [inline docs/Technical-Documentation/circle-ci.md:106]: 😍 Love to see docs getting updated regularly

**Review by andrew-jameson — APPROVED** [review]
LGTM!

File: .circleci/base_config.yml [hunk 4]
```diff
@@ -16,15 +16,18 @@ executors:
       image: ubuntu-2204:2022.10.1
 
 parameters:
+  develop_branch_deploy:
+    type: boolean
+    default: false
   run_dev_deployment:
     type: boolean
     default: false
```
Comment (n0remac) [inline .circleci/base_config.yml:24]: This seems clean and fine, but could you have used `run_dev_deployment` and hooked into the existing `staging-deployment` workflow?

Comment (George-Hudson) [reply .circleci/base_config.yml:24]: I did try that at first, but it seemed to tie in things I did not want as well. I tried adding some logic, but it got pretty messy and still didn't do what I wanted... so I gave up and created a different flag to operate off of.

File: scripts/deploy-backend.sh (hunk 2 above)
Comment (George-Hudson) [reply scripts/deploy-backend.sh:47]: CYPRESS_TOKEN is already in there, yes. It's actually being used to populate the CYPRESS_TOKEN below. DJANGO_DEBUG is also in there, but I'm not trying to grab it, instead just explicitly setting to "yes"

File: tdrs-frontend/package.json (hunk 1 above)
Comment (raftmsohani) [reply tdrs-frontend/package.json]: Nit: we should have all URLs as var. In case we need to change them, we should only have to change in one place

**Review by raftmsohani — APPROVED** [review]

File: tdrs-frontend/package.json (hunk 1 above)
Comment (George-Hudson) [reply tdrs-frontend/package.json]: Converted to ENV VARS and I have updated the env vars.

**Review by ADPennington — APPROVED** [review]
thanks @George-Hudson 👍🏾

**Pull request merged**